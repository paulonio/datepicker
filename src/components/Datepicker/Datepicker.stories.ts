import type { Meta, StoryObj } from '@storybook/react';
import Datepicker from './Datepicker';

const meta: Meta<typeof Datepicker> = {
  title: 'Datepicker',
  component: Datepicker,
};

export default meta;
type Story = StoryObj<typeof Datepicker>;

export const Primary: Story = {
  args: {
    from: new Date(2022, 5, 12),
    to: new Date(2023, 7, 14),
    start: 'su',
    view: 'monthly',
  },
};
