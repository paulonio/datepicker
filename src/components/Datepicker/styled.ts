import styled from 'styled-components';

export const DatepickerWrapper = styled.div`
  width: 720px;
  padding: 100px;
`;
