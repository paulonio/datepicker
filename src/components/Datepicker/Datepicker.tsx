import React, { FC, useRef, useReducer } from 'react';
import { DatepickerWrapper } from './styled';
import Global from '../../styles/Global';
import Input from '../Input/Input';
import Calendar from '../Calendar/Calendar';
import { Mode } from '../../utils/ClassCalendar';
import { useOutsideClick } from '../../hooks/useOutsideClick';

export type View = 'weekly' | 'monthly';

export interface DatepickerProps {
  start: Mode;
  view: View;
  maxDate: Date;
  minDate: Date;
}

export const init: DatepickerProps = {
  start: 'mo',
  view: 'monthly',
  minDate: new Date(2021, 3, 2),
  maxDate: new Date(2025, 4, 25),
};

export interface Init {
  from: Date | null;
  to: Date | null;
  currentCalendar: 'From' | 'To';
  showCalendar: boolean;
}

const initialValue: Init = {
  from: new Date(2022, 5, 17),
  to: new Date(2023, 7, 17),
  currentCalendar: 'From',
  showCalendar: false,
};

export type Action =
  | { type: 'SET_DATE_FROM'; payload: { date: Date | null } }
  | { type: 'SET_DATE_TO'; payload: { date: Date | null } }
  | { type: 'SET_CURRENT_CALENDAR'; payload: { calendar: 'From' | 'To' } }
  | { type: 'TOGGLE_CALENDAR'; payload: { showCalendar: boolean } };

const reducer = (state: Init, action: Action) => {
  switch (action.type) {
    case 'SET_DATE_FROM':
      return { ...state, from: action.payload.date };
    case 'SET_DATE_TO':
      return { ...state, to: action.payload.date };
    case 'SET_CURRENT_CALENDAR':
      return { ...state, currentCalendar: action.payload.calendar };
    case 'TOGGLE_CALENDAR':
      return { ...state, showCalendar: action.payload.showCalendar };
    default:
      return state;
  }
};

const Datepicker: FC<DatepickerProps> = ({ ...initial }) => {
  const [state, dispatch] = useReducer(reducer, initialValue);
  const calendarRef = useRef<HTMLDivElement>(null);

  const handleCloseCalendar = (value: boolean) => {
    dispatch({ type: 'TOGGLE_CALENDAR', payload: { showCalendar: value } });
  };
  useOutsideClick(calendarRef, handleCloseCalendar);

  return (
    <>
      <Global />
      <DatepickerWrapper>
        <Input label="From" date={state.from} dispatch={dispatch} />
        <Input label="To" date={state.to} dispatch={dispatch} />
        {state.showCalendar && (
          <Calendar
            ref={calendarRef}
            config={initial}
            state={state}
            dispatch={dispatch}
            data-component="calendar"
          />
        )}
      </DatepickerWrapper>
    </>
  );
};

export default Datepicker;
