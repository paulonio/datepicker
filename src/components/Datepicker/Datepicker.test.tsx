import React from 'react';
import { render, screen } from '@testing-library/react';
import Datepicker from './Datepicker';

describe('Datepicker', () => {
  const minDate = new Date(2021, 4, 5);
  const maxDate = new Date(2025, 4, 25);

  it('should render datepicker component', () => {
    const { container } = render(
      <Datepicker start="mo" view="monthly" weekend="show" minDate={minDate} maxDate={maxDate} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should display 2 inputs', () => {
    render(
      <Datepicker start="mo" view="monthly" weekend="show" minDate={minDate} maxDate={maxDate} />
    );

    const inputs = screen.getAllByTestId('input');
    expect(inputs).toHaveLength(2);
  });
});
