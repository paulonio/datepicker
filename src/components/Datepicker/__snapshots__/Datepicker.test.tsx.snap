// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker should render datepicker component 1`] = `
<div>
  <div
    class="sc-beqWaB ettSNG"
  >
    <div
      class="sc-gueYoa eiXeJt"
    >
      <div
        class="sc-dmqHEX btCDpK"
      >
        From
      </div>
      <label
        class="sc-hLseeU hiFQXq"
      >
        <div
          class="sc-eDDNvR fmEVhg"
        >
          <img
            class="sc-gLDzan cXjavB"
            src="./assets/icons/Calendar.svg"
          />
        </div>
        <input
          class="sc-iAEyYk vvzef"
          data-testid="input"
          placeholder="Choose Date"
          value=""
        />
      </label>
    </div>
    <div
      class="sc-gueYoa eiXeJt"
    >
      <div
        class="sc-dmqHEX btCDpK"
      >
        To
      </div>
      <label
        class="sc-hLseeU hiFQXq"
      >
        <div
          class="sc-eDDNvR fmEVhg"
        >
          <img
            class="sc-gLDzan cXjavB"
            src="./assets/icons/Calendar.svg"
          />
        </div>
        <input
          class="sc-iAEyYk vvzef"
          data-testid="input"
          placeholder="Choose Date"
          value=""
        />
      </label>
    </div>
  </div>
</div>
`;
